### ISSUE TITLE
binary_cross_entropy_with_logits outputs NaN on -inf, which is inconsistent with documentation

### ISSUE THREAD

**Issue by maybeLee** [pytorch/pytorch#135359]
### 🐛 Describe the bug

When running the following code:

```
import torch
import numpy as np
x = torch.tensor(-np.inf)
y = torch.tensor(0.9)
res1 = torch.nn.functional.binary_cross_entropy_with_logits(x, y)
res2 = torch.nn.BCEWithLogitsLoss()(x, y)
print(res1, res2)  # nan, nan
```
Based on the documentation: https://pytorch.org/docs/stable/generated/torch.nn.BCEWithLogitsLoss.html#torch.nn.BCEWithLogitsLoss.

The calculation of both torch.nn.functional.binary_cross_entropy_with_logits and BCEWithLogitsLoss should be consistent with the following equation:

```
-y(log(sigmoid(x))+(1-y)*log(1-sigmoid(x)))
```

However, the equation outputs `inf` instead of `NaN`:

```
res3 = -(y*torch.log(torch.sigmoid(x)) + (1-y)*torch.log(torch.tensor(1.)-torch.sigmoid(x)))
print(f"The equation's result: {res3}")  # inf
```

After checking some code, I think this might be a numerical instability problem:
https://github.com/pytorch/pytorch/blob/ee1b6804381c57161c477caa380a840a84167676/aten/src/ATen/native/Loss.cpp#L376

The `log_sigmoid_input` will be `-inf` when `x=-inf`. This is because the exponent of the exp operator will be a large positive value. Finally, the computation will involve `0.1*(-np.inf)-(-np.inf)`, thus leading to NaN.

I noticed that TensorFlow made a special handling on large negative input (https://www.tensorflow.org/api_docs/python/tf/nn/sigmoid_cross_entropy_with_logits). 
Specifically, they use the following equation `max(x, 0) - x * z + log(1 + exp(-abs(x)))`, which makes the exponent of the exp operator always negative. It would be nice if torch.nn.functional.binary_cross_entropy_with_logits and BCEWithLogitsLoss can also make such special handling

### Versions

[pip3] numpy==1.26.2
[pip3] optree==0.11.0
[pip3] torch==2.4.0
[pip3] triton==3.0.0
[conda] numpy                     1.26.2                   pypi_0    pypi
[conda] optree                    0.11.0                   pypi_0    pypi
[conda] torch                     2.4.0                    pypi_0    pypi
[conda] triton                    3.0.0                    pypi_0    pypi

**Comment by vadimkantorov** [comment]
Related:
- https://github.com/pytorch/pytorch/issues/31829

**Comment by ezyang** [comment]
Putting triage review in case manufacturing consensus here is easy

**Comment by mikaylagawarecki** [comment]
This is a similar issue to the BC-breaking `safe_softmax` change. The issue caused by landing that change was related to the decomp. We are more hopeful that we can fix without causing issues since this is CompositeExplicit, will followup discussion during nn/optim triage

**Comment by vadimkantorov** [comment]
Also, maybe one way to solving this could be introducing some extra `safe_*`/`nan*`-prefixed functions for the frequent functions which can produce NaNs or inf, but is usually undesired. They might be a bit slower, but in some cases it's okay - and is useful semantics to have. E.g. `safe_div`, `safe_sub`, `safe_mul` more in https://en.wikipedia.org/wiki/NaN#Operations_generating_NaN In NumPy there are a bunch of NaN-ignoring reduction ops like `np.nanmean`. Also, some related old discussion:
- https://github.com/pytorch/pytorch/issues/50122
